Kube-proxy: new bounded frequency runner

aojea opened the pull request:
/kind cleanup
/kind feature



```release-note
NONE
```

Refactored the proxy update scheduling logic by modifying the existing BoundedFrequencyRunner to make it simpler and more suited for kube-proxy use case.

- ensures updates happen only once every minimum interval, coalescing requests within minimal intervals.
- at least once every maximum interval.
- at least once during the retry interval, once a Retry is scheduled, coalescing Retry request and canceling if it ran due to other trigger.


This new runner is a simpler version compared to the previous one, also the burst functionality was removed because its not clear the practical benefit vs the added complexity, we already had the burst size hardcoded to 2 in the proxies, and once the burst tokens are exhausted they require several idle cycles to replenish the burst tokens, that makes only useful for initialization.

It also change the API removing the `Retry()` method and receiving a function that returns an error, making the retry logic based on the error.  The retry gets cancelled if any successful run happens, the `minInterval` is always respected independently of the duration of the `retryInterval`

Tested with `stress` and no flakes 
```
3m35s: 1728 runs so far, 0 failures
3m40s: 1776 runs so far, 0 failures
3m45s: 1824 runs so far, 0 failures
3m50s: 1872 runs so far, 0 failures
...
8m40s: 4272 runs so far, 0 failures
8m45s: 4320 runs so far, 0 failures
8m50s: 4320 runs so far, 0 failures
8m55s: 4368 runs so far, 0 failures
9m0s: 4416 runs so far, 0 failures

```


### Food for thoughts

In the future scheduling algorithms (left for future contributors) we may add different scheduling logic like a self-adaptive one based on execution time, where interval time will be based on x% busy cycle.  E.g. 50% means If it takes 2 seconds, then wait 2 seconds before going again  ( busy_time / idle_time + busy_time) , example strawman approach (without testing)

```go
/*
Copyright 2025 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package runner

import (
	"time"

	"k8s.io/utils/clock"
)

type AdaptativeFrequencyRunner struct {
	maxInterval time.Duration
	busyPercent int
	fn          func()
	queue       chan struct{}

	clock            clock.Clock
	maxIntervalTimer clock.Timer
	idleCycleTimer   clock.Timer
}

// NewAdaptativeFrequencyRunner creates and returns a new AdaptativeFrequencyRunner.
// This runner manages the execution frequency of the provided function `fn` by
// dynamically adjusting the idle time between runs to meet a target busy percentage.
//
// Parameters:
//   - fn: The function to be executed.
//   - busyPercent: The target percentage of time (0-100) that the runner should
//     spend executing `fn`. For example, if busyPercent is 20, the runner aims
//     for the execution time (`busyCycle`) to be 20% of the total cycle time
//     (`busyCycle` + `idleCycle`).
//
// The runner guarantees two properties:
//  1. Adaptive Minimum Interval (`idleCycle`): After `fn` completes (taking
//     `busyCycle` duration), the runner calculates the necessary `idleCycle` using
//     the formula: `idleCycle = busyCycle * (100 - busyPercent) / busyPercent`.
//     It then waits for this `idleCycle` before the *start* of the next execution.
//     Calls to `Run()` during this cooldown period are coalesced and deferred
//     until the calculated interval expires. This prevents burst executions while
//     adapting the cooldown based on the actual execution time.
//     - If `busyPercent` is 0, `idleCycle` effectively becomes infinite (or limited by `maxInterval`).
//     - If `busyPercent` is 100, `idleCycle` becomes 0, causing immediate re-runs.
//  2. Maximum Interval (`maxInterval`): The function `fn` is guaranteed to run
//     at least once per `maxInterval`, ensuring periodic execution even without
//     explicit `Run()` calls (e.g., for refreshing state). Note: The provided code
//     snippet initializes the timer but doesn't set the `maxInterval` value itself;
//     this should be set appropriately for the guarantee to hold.
//
// Internally, the runner uses a buffered channel (capacity 1) as a queue.
// A dedicated goroutine (`Loop`) consumes from this queue or waits for the
// `maxInterval` timer. Calls to `Run()` add to the queue if empty, otherwise,
// they are coalesced.
//
// Panics if `busyPercent` is less than 0 or greater than 100.
func NewAdaptativeFrequencyRunner(fn func(), busyPercent int, maxInterval time.Duration) *AdaptativeFrequencyRunner {
	if busyPercent < 0 || busyPercent > 100 {
		panic("busyPercent must be between 0 and 100")
	}

	bfr := &AdaptativeFrequencyRunner{
		busyPercent: busyPercent,
		maxInterval: maxInterval,
		queue:       make(chan struct{}, 1),
		fn:          fn,
		clock:       clock.RealClock{},
	}
	return bfr
}

func (bfr *AdaptativeFrequencyRunner) Loop(stop <-chan struct{}) {
	defer close(bfr.queue)
	defer bfr.idleCycleTimer.Stop()

	bfr.maxIntervalTimer = bfr.clock.NewTimer(bfr.maxInterval)
	for {
		select {
		case <-stop:
			return
		case <-bfr.maxIntervalTimer.C(): // process work at least once every maxInterval
		case <-bfr.queue: // process scheduled work
		}

		// execute
		start := bfr.clock.Now()
		bfr.fn()
		busyCycle := bfr.clock.Since(start)

		interval := time.Duration(float64(busyCycle) * float64(100-bfr.busyPercent) / float64(bfr.busyPercent))

		// Reschedule for next execution
		bfr.maxIntervalTimer.Stop()
		bfr.maxIntervalTimer.Reset(bfr.maxInterval)
		bfr.idleCycleTimer.Stop()
		bfr.idleCycleTimer.Reset(interval)

		select {
		case <-stop:
			return
		case <-bfr.idleCycleTimer.C(): // wait at least for idleCycleTimer until next execution
		}
	}
}

// Run requests an execution of the configured function `fn`.
func (bfr *AdaptativeFrequencyRunner) Run() {
	// It attempts to send a signal to the internal queue. If the queue is already
	// full (meaning an execution is already pending or waiting for the idleCycle),
	// this call is effectively coalesced with the previous one, and no new signal
	// is added. This ensures that rapid calls to Run() don't queue up indefinitely.
	select {
	case bfr.queue <- struct{}{}:
	default:
	}
}
```

aojea commented:
/assign @danwinship @thockin  @aroradaman

toVersus commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -0,0 +1,126 @@
+		}
+
+		// execute
+		bfr.fn()
```
Is it possible for two executions to occur if `Run()` is called simultaneously with `maxInterval`?

aojea replied:
golang `select` statement guarantees that only one wins, then the next execution will be wait for the `minSyncPeriod`

aojea commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -0,0 +1,126 @@
+// this function will panic.
+func NewBoundedFrequencyRunner(fn func(), minInterval, maxInterval time.Duration) *BoundedFrequencyRunner {
+	if maxInterval < minInterval {
+		panic("maxInterval must be greater or equal than minInterval")
```
read again, it panics  `maxInterval` is lower than minInterval

toVersus replied:
Ah, sorry, I get it now🤦‍♂️

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -818,7 +818,7 @@ func (proxier *Proxier) syncProxyRules() {
 		if !success {
 			proxier.logger.Info("Sync failed", "retryingTime", proxier.syncPeriod)
-			proxier.syncRunner.RetryAfter(proxier.syncPeriod)
+			proxier.syncRunner.Run()
```
I think you're confusing `syncPeriod` and `minSyncPeriod`.

The goal is to try again but specifically _not_ to try again as soon as possible, since this failure is supposed to be a "can't happen" thing, so we assume if we try again immediately we might hit the same error again. Especially, in the iptables case, if someone else is holding the xtables lock for a really long time and we time out trying to acquire it, we want to wait and give them time to finish what they were doing.

danwinship replied:
Yes. `RetryAfter(x)` effectively means "if `x` elapses without a sync happening, then sync". (The same way that the construct-time `maxInterval` means "if `maxInterval` elapses without a sync happening, then sync".)

`RetryAfter` was added a while after the rest of BFR, so it's not as well-integrated with the code as it might have been. (It's also tricky because we wanted to make sure that it works both from inside and from outside a sync, which made the locking tricky.)

danwinship commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -0,0 +1,128 @@
+//
+// `maxInterval` must be greater than or equal to `minInterval`; otherwise,
+// this function will panic.
+func NewBoundedFrequencyRunner(fn func(), minInterval, maxInterval time.Duration) *BoundedFrequencyRunner {
```
It would be helpful to have one commit that just copies BFR over to `pkg/proxy/runner` unchanged, then a second commit to get to this state, to make it clearer what is the same and what is different from the old code.

danwinship commented on pkg/proxy/runner/interface.go:
```diff
@@ -0,0 +1,26 @@
+// specific algorithm implemented by the concrete type. Implementations typically
+// manage the timing and frequency of runs.
+// To start the scheduling loop, call the Loop() method in a separate goroutine.
+type Runner interface {
```
I'm not sure this generic interface is useful.

In particular, as discussed in the other PR, if we are moving BFR to `pkg/proxy`, then we should absolutely make it more proxy-specific. Especially, keeping explicit track of partial syncs and full syncs as separate things with not-entirely-overlapping timers, rather than having the proxies themselves keep track of which kind of sync they wanted to do when the single BFR timer triggers.

At any rate, if people want to implement other kinds of runners in the future, we can add an interface then.

aojea replied:
fine, it was related to an offline conversation with Tim, where we was suggesting to move to a more adaptative runner based on cpu busy time, see description for more details ... I didn't want to mix both things but I also wanted to use the review to get feedback, I will refactor the PR accordingly

danwinship replied:
I think that may also fit in with what I was saying about "more proxy-specific". Presumably if we wanted "adaptive based on CPU busy time" we'd want it for all proxy backends, so we could just modify the single runner implementation rather than needing a second implementation, right?

(I bet Tim didn't realize that BoundedFrequencyRunner is currently used _only_ by kube-proxy for scheduling syncs, so there's no real need to worry about "what kube-proxy wants from BFR" vs "what other people might want from BFR" since there aren't currently any other people.) (And, as per #130518 it probably makes more sense for anything that isn't already using BFR to use workqueues instead.)

aojea replied:
👍

aojea committed:
move async proxyrunner to kube-proxy only

aojea commented:
/assign @danwinship @thockin

thockin commented:
This is interesting but I have no capacity to review it in the near term.  I trust you and Dan to work it out, and I can final approve the final change

danwinship commented:
commit message needs to be updated...

danwinship commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -21,188 +21,141 @@ import (
+	defer bfr.minIntervalTimer.Stop()
+
+	bfr.maxIntervalTimer = bfr.clock.NewTimer(bfr.maxInterval)
+	defer bfr.maxIntervalTimer.Stop()
```
couldn't you create the `retryTimer` here too to be more consistent?

aojea replied:
I did, but the problem is that it panics on the kube-proxy unit tests that does not start the loop but call retry ... so there was the option of running the loop on the kube-proxy unit test or initialize the stopped timer at creation time, I choose the later for simplicity to avoid having to start this runner on the unit tests

aojea replied:
fixed, we no longer use `Retry()`

danwinship commented on pkg/proxy/runner/bounded_frequency_runner_test.go:
```diff
@@ -17,437 +17,474 @@ limitations under the License.
+		time.Sleep(10 * time.Millisecond)
+	}
+	return false
 }
```
`waitForNCalls`?

Also, it's very confusing that this is sleeping _real_ time as opposed to FakeClock time... if we are keeping it like this, that needs to be made much more explicit.

aroradaman replied:
I think we do need some real sleep for context switching here
Tried to simplify the `receiver` 

```diff
diff --git a/pkg/proxy/runner/bounded_frequency_runner_test.go b/pkg/proxy/runner/bounded_frequency_runner_test.go
index 92f0eb9e2c8..bd484a1dd65 100644
--- a/pkg/proxy/runner/bounded_frequency_runner_test.go
+++ b/pkg/proxy/runner/bounded_frequency_runner_test.go
@@ -17,7 +17,6 @@ limitations under the License.
 package runner
 
 import (
-	"sync/atomic"
 	"testing"
 	"time"
 
@@ -26,106 +25,98 @@ import (
 
 // Track calls to the managed function.
 type receiver struct {
-	counter atomic.Int32
+	counter chan struct{}
 }
 
 func (r *receiver) F() {
-	r.counter.Add(1)
+	r.counter <- struct{}{}
 }
 
+// calls returns the number of times F() has been called,
+// it waits for a reasonable amount of time (10ms) for context
+// switching between different go routines by the go scheduler
 func (r *receiver) calls() int {
-	return int(r.counter.Load())
-}
-
-func (r *receiver) waitForValue(n int) bool {
-	for i := 0; i < 10; i++ {
-		if r.calls() == n {
-			return true
-		}
-		time.Sleep(10 * time.Millisecond)
-	}
-	return false
+	time.Sleep(10 * time.Millisecond)
+	return len(r.counter)
 }
 
 func Test_BoundedFrequencyRunner(t *testing.T) {
 	var minInterval = 1 * time.Second
 	var retryInterval = 5 * time.Second
 	var maxInterval = 10 * time.Second
-	obj := &receiver{}
+	obj := &receiver{counter: make(chan struct{}, 100)}
 	fakeClock := clock.NewFakeClock(time.Now())
 	runner := construct("test-runner", obj.F, minInterval, retryInterval, maxInterval, fakeClock)
 	stop := make(chan struct{})
 	defer close(stop)
 
 	go runner.Loop(stop)
-	// Wait for the loop to be running.
-	time.Sleep(20 * time.Millisecond)
 
 	// Run once, immediately.
 	// rel=0ms
 	runner.Run()
-	if !obj.waitForValue(1) {
+	if obj.calls() != 1 {
 		t.Errorf("expected one call, got %d", obj.calls())
 	}

```

aojea replied:
this is same feedback as in https://github.com/kubernetes/kubernetes/pull/131502#discussion_r2069479576

aojea replied:
your suggestion Daman is really good, I think your comment is spot on, let me try with channels

aojea replied:
see new approach based on current feedback

danwinship commented on pkg/proxy/runner/bounded_frequency_runner_test.go:
```diff
@@ -17,437 +17,474 @@ limitations under the License.
+	runner.Run()
+	if !obj.waitForValue(1) {
+		t.Errorf("expected one call, got %d", obj.calls())
+	}
```
The name "waitForValue" is confusing here since what you really mean is "wait a bit and confirm that the value didn't change".

I don't like the way this unit test works. I feel like it would be better to change how the runner works internally so that the unit test can be completely race-free.

aojea replied:
the problem is that I reused the helper and the name is bad, for fakeclock test I should not use this, as you don't need to active wait for the call since we control the clock steps ... I was trying to keep the diff small but I make it more confusing

danwinship commented on pkg/proxy/runner/bounded_frequency_runner_test.go:
```diff
@@ -17,437 +17,474 @@ limitations under the License.
-	upd = <-timer.updated // wait for reset
-	checkTimer(name, t, upd, true, expectNext)
-}
+func Test_BoundedFrequencyRunnerRetry_RealClock(t *testing.T) {
```
Why? The whole point of using `clock.Clock` is so we can avoid needing to do racy tests with the real clock.

aojea replied:
because fakeclock are not reals and are not  a one to one replacement of a real clock, I would not test something only using fake clocks, that will not give you real coverage

danwinship replied:
how are they not real? The _entire point_ of FakeClock is to be usable as a replacement for the real clock in unit tests

aojea replied:
well, I can assure you they are not, is fake, so it is as close and good as the people implemented it ... with all the possible bugs and limitations, check the implementation if you are interested ... real clocks are much more complex, I went with clayton deep on this https://github.com/kubernetes/kubernetes/pull/107826  ... fake clock is a mock, and as a mock it works well, but is not a replacement ...

aroradaman commented on pkg/proxy/runner/bounded_frequency_runner_test.go:
```diff
@@ -17,437 +17,474 @@ limitations under the License.
-type fakeTimer struct {
-	c chan time.Time
+	go runner.Loop(stop)
+	// Wait for the loop to be running.
```
Since we are already restructuring BFR can we have something like `runner.StartLoop()` which internally starts the go-routine and ensures that it has started before returning.

aojea replied:
yep, we need to make this synchronous

aojea replied:
added a channel to synchronize readiness

aojea commented on pkg/proxy/runner/bounded_frequency_runner_test.go:
```diff
@@ -17,437 +17,474 @@ limitations under the License.
-	active bool
-	next   time.Duration // iff active == true
-}
+func Test_BoundedFrequencyRunner(t *testing.T) {
```
ack

aojea commented on pkg/proxy/runner/bounded_frequency_runner_test.go:
```diff
@@ -17,437 +17,474 @@ limitations under the License.
-type fakeTimer struct {
-	c chan time.Time
+	go runner.Loop(stop)
+	// Wait for the loop to be running.
+	time.Sleep(20 * time.Millisecond)
```
yep, we should make this synchronous , I will change this once we agree on the overall idea

aojea commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -213,101 +166,26 @@ func (bfr *BoundedFrequencyRunner) Loop(stop <-chan struct{}) {
 	bfr.retryMu.Lock()
 	defer bfr.retryMu.Unlock()
-	if !bfr.retryTime.IsZero() && bfr.retryTime.Before(retryTime) {
+	if bfr.retrying {
```
I don't know if there are performance implications on mutex vs select, both are synchronous primitives, but this runner will not get any impact from one or the other ... the code just replaces the bool by the channel , feel the bool is easier to read ... what we really need is a delayed queue, but this complex to implement , see the workqueue implementation

aojea replied:
we need to lock the retry fields or there can be some races during the function execution

aojea commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -21,188 +21,141 @@ import (
+// Internally, the runner uses a buffered channel with a capacity of 1, acting
+// as a single-item queue. A dedicated goroutine (started by `Loop`) consumes
+// from this queue. This consumer loop triggers an execution of `fn` under two
+// conditions:
```
ack

aojea commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -21,188 +21,141 @@ import (
-	<-timer.C()                                  // consume the first tick
-	return construct(name, fn, minInterval, maxInterval, burstRuns, timer)
+// `maxInterval` must be greater than or equal to `minInterval`; otherwise,
+// this function will panic.
```
if retryInterval  > maxInterval it will work, because maxInterval will trigger and always disarm the timer.
if retryInterval < minInterval, it will always have to wait for minInterval until next execution

The select statement guarantees that only one of the three conditions triggers a run, and after an execution, state is restored and minInterval has to pass

aojea replied:
it always wait for minInterval, I added an unit test for that, the Retry logic is locked during execution and can not race with the other triggers, and after a run it ALWAYS wait for minInterval

aojea commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -21,188 +21,141 @@ import (
 func (bfr *BoundedFrequencyRunner) Loop(stop <-chan struct{}) {
-	klog.V(3).Infof("%s Loop running", bfr.name)
-	bfr.timer.Reset(bfr.maxInterval)
+	klog.V(3).Infof("%s Loop running", bfr.name) // nolint:logcheck
```
done

aojea commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -21,188 +21,141 @@ import (
-			bfr.stop()
-			klog.V(3).Infof("%s Loop stopping", bfr.name)
+			klog.V(3).Infof("%s Loop stopping", bfr.name) // nolint:logcheck
 			return
```
yep

aojea replied:
added the defer for this timer too

aojea committed:
add kube-proxy BoundedFrequencyRunner implementation

This implementation replaces an older, more complex BoundedFrequencyRunner
that included burst capabilities based on a token bucket algorithm. Bursts
may be only useful during bootstrap, bear in mind that after the burst
token are exhaused then it will require several min intervals to be
replenished. Since the bursts value were already hardcoded in the
proxies to 2 and there is not evidence of its utility, this runners
sacrifies bursts for simplicity and provide the following
characteristics:

1.  Runs at most once per `minInterval`.
2.  Runs at least once per `maxInterval`.
3.  A retry is scheduled using retryInterval only if fn returns a
non-nil error. If fn succeeds (returns nil), any pending retry
timer is stopped.

Change-Id: I8378d8c67e5222f6c4e5b34d76770996475caf4f

aojea commented on pkg/proxy/runner/bounded_frequency_runner_test.go:
```diff
@@ -17,437 +17,412 @@ limitations under the License.
+	// probability that the maxInterval didn't get reset before the flakeClock advanced
+	// > 1824 runs so far, 1 failures (0.05%)
+	// add a sleep to handle that edge case until fake clock allow to expose the number of waiters
+	time.Sleep(1 * time.Millisecond)
```
this will be fixed with https://github.com/kubernetes/utils/pull/325

danwinship commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -21,188 +21,143 @@ import (
+
+			// do not allow to reset the retryInterval during the execution
+			bfr.retryMu.Lock()
+			defer bfr.retryMu.Unlock()
```
The only time we ever call Retry() is from within the execution of fn().

danwinship replied:
As an alternative API, we could get rid of Retry() and have `fn()` return a value indicating if it wanted a retry

aojea replied:
that sounds an interesting idea

aojea replied:
@danwinship see https://github.com/kubernetes/kubernetes/pull/131502/commits/b6f091534ff20dddf236baf132f9af5242f78e3a for strawman proposal it looks pretty neat and simple, also the diff is very small if we use a named variable on the proxies , If you give me ack on that I will implement it that way in a new commit

danwinship commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -136,18 +131,17 @@ func (bfr *BoundedFrequencyRunner) Loop(stop <-chan struct{}) {
+				klog.V(3).InfoS("retrying", "name", bfr.name, "interval", bfr.retryInterval)
+				bfr.retryIntervalTimer.Reset(bfr.retryInterval)
+			} else {
+				bfr.retryIntervalTimer.Stop()
```
You shouldn't need a separate `retryIntervalTimer`; you can just reset `maxIntervalTimer` to either `retryInterval` or `maxInterval` depending on whether there was an error

aojea replied:
good point

aojea replied:
fixed in next commit

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -818,7 +817,7 @@ func (proxier *Proxier) syncProxyRules() {
 		if !success {
 			proxier.logger.Info("Sync failed", "retryingTime", proxier.syncPeriod)
-			proxier.syncRunner.RetryAfter(proxier.syncPeriod)
+			proxier.syncRunner.Retry()
```
I'd merge the retryErr change in with this commit, since this change is incorrect as written (it will deadlock on `retryMu`).

aojea replied:
done

danwinship reviewed (commented):
maybe split the utils bump out?

danwinship commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -136,18 +131,17 @@ func (bfr *BoundedFrequencyRunner) Loop(stop <-chan struct{}) {
+			err := bfr.fn()
+			// an error will schedule a retry after the retryInterval,
+			// any successful run until that period will stop the retry attempt.
+			if err != nil {
```
Maybe have a global `var RetryError` and only retry if `errors.Is(err, RetryError)`?

aojea replied:
we are not going to expose this and the usage is mostly by historical reasons, do we plan to do something if the error is not retryable?

aojea committed:
use a single timer for max and retry intervals

Change-Id: I3a3307c6a883a29664e12028977c82b95cae5565

aojea commented on pkg/proxy/runner/bounded_frequency_runner.go:
```diff
@@ -136,18 +131,17 @@ func (bfr *BoundedFrequencyRunner) Loop(stop <-chan struct{}) {
+			// an error will schedule a retry after the retryInterval,
+			// any successful run until that period will stop the retry attempt.
+			if err != nil {
+				klog.V(3).InfoS("retrying", "name", bfr.name, "interval", bfr.retryInterval)
```
done

aojea commented:
> {Failed  === RUN   TestCancelEviction/slice-deleted

 known flake https://github.com/kubernetes/kubernetes/issues/131072

/test pull-kubernetes-unit

aojea commented:
/test pull-kubernetes-unit

there are no nodes running

danwinship commented:
see  #131615; feel free to either close this in favor of that or pull commits from that into this

aojea commented:
> see #131615; feel free to either close this in favor of that or pull commits from that into this

/close

superseded by https://github.com/kubernetes/kubernetes/pull/131615
